### PR TITLE
cxx-qt-gen: use impl cxx_qt::QObject<T> block for invokables

### PR DESCRIPTION
**Pull request by ahayzen-kdab**
This helps later as the invokables are implemented onto a different
struct, so this clearly defines the split between C++ and Rust
methods etc.

### CODE REVIEW

**Comment by ahayzen-kdab** [comment]
@LeonMatthesKDAB  did we decide if we were going to call this `impl Qt<T>` or `impl cxx_qt<T>` ? As i think the `Qt` felt more natural for saying these are "Qt" invokables, but the `cxx_qt` was better to avoid clashes ? Maybe even `cxx_qt::invokables` could be considered as that then matches `cxx_qt::signals` etc ?

File: book/src/qobject/rustobj_struct.md [hunk 1]
```diff
@@ -20,12 +20,12 @@ The RustObj struct allows you to define the following items
 
 ## Invokables
 
-To define a method which is exposed to QML and C++, add a method on the `RustObj` struct and add the attribute `#[invokable]`. The parameters and return type are then matched to the Qt side. Also CXX-Qt automatically adds wrapper code around your invokable to automatically perform any conversion between the [C++ and Rust types](../concepts/types.md).
+To define a method which is exposed to QML and C++, add a method on the `RustObj` struct, with `impl Qt<RustObj>`, and add the attribute `#[invokable]`. The parameters and return type are then matched to the Qt side. Also CXX-Qt automatically adds wrapper code around your invokable to automatically perform any conversion between the [C++ and Rust types](../concepts/types.md).
```
Comment (Be-ing) [inline book/src/qobject/rustobj_struct.md:23]:
```suggestion
To define a method which is exposed to QML and C++, add a method on the `RustObj` struct, with `impl Qt<RustObj>`, and add the attribute `#[invokable]`. The parameters and return type are then matched to the Qt side. Also CXX-Qt adds wrapper code around your invokable to automatically perform any conversion between the [C++ and Rust types](../concepts/types.md).
```

`automatically` is repeated

File: cxx-qt-gen/src/extract.rs [hunk 2]
```diff
@@ -960,16 +951,20 @@ pub fn extract_qobject(
...
-                                let mut extracted = extract_invokables(
-                                    &original_impl.items,
-                                    cpp_namespace_prefix,
-                                    &qt_ident,
-                                )?;
-
-                                object_invokables.append(&mut extracted.invokables);
-                                object_normal_methods.append(&mut extracted.normal_methods);
+                                // Invokables aren't on the impl T block
+                                object_normal_methods.append(&mut original_impl.items);
```
Comment (Be-ing) [inline cxx-qt-gen/src/extract.rs:955]: Should using `#[invokable]` outside of `impl Qt<RustObj>` cause an error?

Comment (ahayzen-kdab) [reply cxx-qt-gen/src/extract.rs:955]: I think it fails to build as there is no macro called `invokable` :-)

**Comment by ahayzen-kdab** [comment]
> @LeonMatthesKDAB did we decide if we were going to call this `impl Qt<T>` or `impl cxx_qt<T>` ? As i think the `Qt` felt more natural for saying these are "Qt" invokables, but the `cxx_qt` was better to avoid clashes ? Maybe even `cxx_qt::invokables` could be considered as that then matches `cxx_qt::signals` etc ?

So i'm wondering if this is the route (assuming valid syntax)

```rust
impl cxx_qt::invokables<MyObject> {
  #[invokable]
  pub fn my_invokable(self: Pin<&mut Self>) {

  }
}

impl MyObject {
  pub fn rust_only_method(&self) {

  }
}
```

@LeonMatthesKDAB where did our discussion end up on having like `#[invokable(return-as-value)]` and `#[property(as-value)]` stuff for `UniquePtr` end up? I can't see it now as there were options like `#[invokable(return QColor)]` as well...

**Comment by LeonMatthesKDAB** [comment]
> So i'm wondering if this is the route (assuming valid syntax)
> ```
> impl cxx_qt::invokables<MyObject> {
>   #[invokable]
>   pub fn my_invokable(self: Pin<&mut Self>) {
> 
>   }
> }
> 
> impl MyObject {
>   pub fn rust_only_method(&self) {
> 
>   }
> }
> ```
We shouldn't use `impl invokables<MyObject>`, as it still makes sense to have non-invokable (aka not to C++ exposed) methods defined within the `impl Qt<MyObject>` (or whatever we end up calling it).
Whilst you can define Rust-only methods directly on MyObject, you might want to have a private method that is still capable of emitting a Signal for example. That you would implement on `Qt<MyObject>` but without `#[invokable]`.

But regarding the `cxx_qt::` prefix. That does make sense for consistencies sake.
So something like: `impl cxx_qt::QObject<MyObject>` or `impl cxx_qt::Qt<MyObject>`. My local playground project even uses `impl cxx_qt<MyObject>`, but maybe `impl cxx_qt::Qt<MyObject>` is a bit more consistent with the other macros we have.
Later we could add the ability to add a `use cxx_qt::Qt` or `use cxx_qt::prelude::*`, to reduce the boilerplate.

> @LeonMatthesKDAB where did our discussion end up on having like #[invokable(return-as-value)] and #[property(as-value)] stuff for UniquePtr end up? I can't see it now as there were options like #[invokable(return QColor)] as well...
> 
I'm unsure whether we reached a consensus on that or whether we were still just discussing options :thinking: 
Do you need a decision on this for this PR already?

File: examples/qml_with_threaded_logic/src/lib.rs [hunk 3]
```diff
@@ -124,6 +115,17 @@ mod website {
         }
     }
 
+    impl RustObj {
+        fn process_event(&mut self, event: &Event, cpp: &mut CppObj) {
```
Comment (LeonMatthesKDAB) [inline examples/qml_with_threaded_logic/src/lib.rs:119]: This needs to be in the `impl Qt<MyObject>` block, not in the `impl RustObj` block, as we need access to the C++ functionality.

File: cxx-qt-gen/src/extract.rs [hunk 4]
```diff
@@ -417,23 +417,13 @@ fn extract_type_ident(
     })
 }
 
-/// The result of extracting invokables from an impl block.
-/// Only intended for internal use for now.
-struct ExtractedInvokables {
-    /// Impl methods that will also be exposed as invokables to Qt
-    invokables: Vec<Invokable>,
-    /// Impl methods that will only be visible on the Rust side
-    normal_methods: Vec<ImplItemMethod>,
```
Comment (LeonMatthesKDAB) [inline cxx-qt-gen/src/extract.rs:426]: As mentioned, we still need these `normal_methods`.

File: book/src/qobject/rustobj_struct.md [hunk 5]
```diff
@@ -20,12 +20,12 @@ The RustObj struct allows you to define the following items
 
 ## Invokables
 
-To define a method which is exposed to QML and C++, add a method on the `RustObj` struct and add the attribute `#[invokable]`. The parameters and return type are then matched to the Qt side. Also CXX-Qt automatically adds wrapper code around your invokable to automatically perform any conversion between the [C++ and Rust types](../concepts/types.md).
+To define a method which is exposed to QML and C++, add a method on the `RustObj` struct, with `impl Qt<RustObj>`, and add the attribute `#[invokable]`. The parameters and return type are then matched to the Qt side. Also CXX-Qt adds wrapper code around your invokable to automatically perform any conversion between the [C++ and Rust types](../concepts/types.md).
```
Comment (LeonMatthesKDAB) [inline book/src/qobject/rustobj_struct.md:23]: Maybe split this up into two paragraphs:
1. What is `impl Qt<RustObj>` (aka, implementation on the C++ QObject that wraps the RustObj - It has access to both C++ and Rust methods).
2. How can I define an invokable on it (#[invokable])

File: book/src/qobject/rustobj_struct.md [hunk 6]
```diff
@@ -20,12 +20,12 @@ The RustObj struct allows you to define the following items
...
 
 Note to access properties on the C++ object use [Cpp Object](./cpp_object.md).
 
 ## Private Methods and Fields
 
 Unlike the [Data Struct](./data_struct.md) fields which are defined on the `RustObj` struct are not exposed as properties to Qt. These can be considered as "private to Rust" fields, and are useful for storing channels for threading or internal information for the QObject.
 
-Methods implemented on the `RustObj` that do not have an `#[invokable]` attribute are not exposed to C++ and are considered "private to Rust" methods. Similar to fields these are useful for threading and internal information.
+Methods implemented on the `RustObj` (and not the `impl Qt<RustObj>`) that do not have an `#[invokable]` attribute are not exposed to C++ and are considered "private to Rust" methods. Similar to fields these are useful for threading and internal information.
```
Comment (LeonMatthesKDAB) [inline book/src/qobject/rustobj_struct.md:31]: Maybe rephrase this a bit to focus on the access to C++ features.
> Methods implemented using `impl RustObj` (and not `impl Qt<RustObj>`) are just normal Rust member methods.
> Therefore they do not have access to any C++/QObject functionality (e.g. emitting Signals, changing properties, etc.)
> You will usually only need to use `impl RustObj` if you want to also use your `RustObj` struct as a normal Rust struct, that is not wrapped in a QObject.

**Review by LeonMatthesKDAB — CHANGES_REQUESTED** [review]
Going in the right direction. A bit of clarification regarding the purposes of `impl Qt<...>` vs `impl ...` are still necessary :)

File: examples/qml_with_threaded_logic/src/lib.rs (hunk 3 above)
Comment (ahayzen-kdab) [reply examples/qml_with_threaded_logic/src/lib.rs:119]: Actually this one is a rust private method (as Event isn't a supported type), what happens is that the handle_update_request calls this method and passes in CppObj. But i've added support for both anyway.

Comment (LeonMatthesKDAB) [reply examples/qml_with_threaded_logic/src/lib.rs:119]: Yes, but it will in future still need to be implemented in the `impl MyObjectQt` block (e.g. `impl cxx_qt::QObject<T>`).

That doesn't mean it needs to be put in the CXX bridge, so it will still be rust-private.

File: book/src/qobject/rustobj_struct.md (hunk 5 above)
Comment (ahayzen-kdab) [reply book/src/qobject/rustobj_struct.md]: Done

File: book/src/qobject/rustobj_struct.md (hunk 6 above)
Comment (ahayzen-kdab) [reply book/src/qobject/rustobj_struct.md]: Done

File: cxx-qt-gen/src/extract.rs (hunk 4 above)
Comment (ahayzen-kdab) [reply cxx-qt-gen/src/extract.rs:426]: Done

File: cxx-qt-gen/src/extract.rs [hunk 7]
```diff
@@ -1008,16 +1014,24 @@ pub fn extract_qobject(
...
-                                object_invokables.append(&mut extracted.invokables);
-                                object_normal_methods.append(&mut extracted.normal_methods);
+                                // Rust only methods are on the impl T block
+                                object_rust_methods.append(&mut original_impl.items);
                             }
                         }
+                        // Invokables are defined in the impl Qt<T> block
+                        "cxx_qt"
+                            if path.segments.len() > 1
+                                && path.segments[1].ident.to_string().as_str() == "QObject" =>
```
Comment (LeonMatthesKDAB) [inline cxx-qt-gen/src/extract.rs:1024]: The premature optimizer monkey in my brain tells me this can probably be "simplified":
`path.segments.get(1).filter(|seg| seg.ident.to_string().as_str() == "QObject").is_some()`
Then we don't need the segments.len() check ;)

Anyway, interesting mental exercise, not important

File: cxx-qt-gen/test_outputs/invokables.h [hunk 8]
```diff
@@ -36,6 +36,11 @@ class MyObject : public QObject
   Q_INVOKABLE qint32 invokableReturnPrimitive();
   Q_INVOKABLE QString invokableReturnStatic();
 
+  void cppOnlyMethod();
```
Comment (LeonMatthesKDAB) [inline cxx-qt-gen/test_outputs/invokables.h:39]: These 4 methods should actually not be generated into C++.
They just need to be in the `impl cxx_qt::QObject<T>` block because that's the only place, where, in future, you'll still have access to C++ functionality (e.g. signals, etc.).

File: cxx-qt-gen/test_outputs/invokables.rs [hunk 9]
```diff
@@ -156,8 +165,33 @@ mod cxx_qt_my_object {
             QString::from_str("static")
         }
 
+        pub fn cpp_only_method_cpp_obj_wrapper(&mut self, cpp: std::pin::Pin<&mut FFICppObj>) {
```
Comment (LeonMatthesKDAB) [inline cxx-qt-gen/test_outputs/invokables.rs:168]: We shouldn't generate a wrapper for this.
It should just be 1-to-1 translated over into the `impl RustObj` block, and later the `impl MyObjectQt` block.

File: cxx-qt/src/lib.rs [hunk 10]
```diff
@@ -60,6 +64,38 @@ pub fn signals(_args: TokenStream, _input: TokenStream) -> TokenStream {
...
+///         #[invokable]
+///         fn invokable(&self, a: i32, b: i32) -> i32 {
+///             a + b
+///         }
+///     }
+/// }
+/// ```
+#[proc_macro]
+#[allow(non_snake_case)]
+pub fn QObject(_input: TokenStream) -> TokenStream {
```
Comment (LeonMatthesKDAB) [inline cxx-qt/src/lib.rs:95]: This shouldn't be a macro, it needs to be a type:
``` rust
pub struct QObject<T> {
    phantom: std::marker::PhantomData<T>
}
```

File: examples/qml_with_threaded_logic/src/lib.rs (hunk 3 above)
Comment (LeonMatthesKDAB) [reply examples/qml_with_threaded_logic/src/lib.rs:125]: In my opinion, this still needs to be in `impl cxx_qt::QObject<RustObj>`, as it modifies a C++ property (e.g. `cpp.set_title`).

**Review by LeonMatthesKDAB — CHANGES_REQUESTED** [review]

File: cxx-qt/src/lib.rs (hunk 10 above)
Comment (ahayzen-kdab) [reply cxx-qt/src/lib.rs:95]: We can't export non macro things in a proc-macro crate, so it has to be this.

File: cxx-qt-gen/test_outputs/invokables.rs (hunk 9 above)
Comment (ahayzen-kdab) [reply cxx-qt-gen/test_outputs/invokables.rs:168]: Eventually yup, this is just the first step, still using the existing wrapper generation. As this impl cxx_qt::QObject block is not implemented on the C++ object yet.

File: cxx-qt-gen/test_outputs/invokables.h (hunk 8 above)
Comment (ahayzen-kdab) [reply cxx-qt-gen/test_outputs/invokables.h:39]: Eventually yes, when we move the implementation of the invokables onto the C++ object. For now we still have the proxies.

File: cxx-qt-gen/test_outputs/invokables.rs (hunk 9 above)
Comment (ahayzen-kdab) [reply cxx-qt-gen/test_outputs/invokables.rs:168]: OK, as discussed i see what you mean now, we need these only on the Rust side, but on the C++ object later.

File: cxx-qt-gen/test_outputs/invokables.h (hunk 8 above)
Comment (ahayzen-kdab) [reply cxx-qt-gen/test_outputs/invokables.h:39]: Removed :-)

File: book/src/qobject/rustobj_struct.md [hunk 11]
```diff
@@ -20,12 +20,17 @@ The RustObj struct allows you to define the following items
 
 ## Invokables
 
-To define a method which is exposed to QML and C++, add a method on the `RustObj` struct and add the attribute `#[invokable]`. The parameters and return type are then matched to the Qt side. Also CXX-Qt automatically adds wrapper code around your invokable to automatically perform any conversion between the [C++ and Rust types](../concepts/types.md).
+A `impl cxx_qt::QObject<RustObj>` is used to define invokables, the `impl cxx_qt::QObject<RustObj>` defines that the methods are implemented onto the C++ QObject.
+Therefore they have access to both C++ and Rust methods. Also CXX-Qt adds wrapper code to around your invokables to automatically perform any conversion between the [C++ and Rust types](../concepts/types.md).
```
Comment (LeonMatthesKDAB) [inline book/src/qobject/rustobj_struct.md:24]: `wrapper code to around` => `wrapper code around`

**Review by LeonMatthesKDAB — COMMENTED** [review]
Just one typo, then we're good to go :)

File: book/src/qobject/rustobj_struct.md (hunk 11 above)
Comment (ahayzen-kdab) [reply book/src/qobject/rustobj_struct.md]: Fixed

**Review by LeonMatthesKDAB — APPROVED** [review]
LGTM
